From 3eca506f989ab934ef9bafeb6e9238ff89006eab Mon Sep 17 00:00:00 2001
From: Adam Borowski <kilobyte@angband.pl>
Date: Sat, 3 Dec 2016 09:00:46 +0100
Subject: [PATCH 1/2] Set "word chars" to gpm's default, hard-coded.

This means alphanumerics, latin-1 letters, _, and, unlike kernel's
default, also -./
---
 src/consolation.h |  1 +
 src/input.c       |  2 ++
 src/selection.c   | 26 ++++++++++++++++++++++++++
 3 files changed, 29 insertions(+)

diff --git a/src/consolation.h b/src/consolation.h
index 052a1a0..cf74711 100644
--- a/src/consolation.h
+++ b/src/consolation.h
@@ -29,6 +29,7 @@ void select_word(int x, int y);
 void select_line(int x, int y);
 void paste(void);
 void scroll(int sc);
+void set_lut(void);
 
 /* action.c */
 
diff --git a/src/input.c b/src/input.c
index 74b271a..846d585 100644
--- a/src/input.c
+++ b/src/input.c
@@ -227,6 +227,8 @@ event_main(void)
   if (!li)
     return 1;
 
+  set_lut();
+
   mainloop(li);
 
   libinput_unref(li);
diff --git a/src/selection.c b/src/selection.c
index cdebb64..912dd0b 100644
--- a/src/selection.c
+++ b/src/selection.c
@@ -20,6 +20,7 @@
 #include <fcntl.h>
 #include <sys/ioctl.h>
 #include <linux/tiocl.h>
+#include <stdint.h>
 
 #include "consolation.h"
 
@@ -108,3 +109,28 @@ void scroll(int sc)
     perror("scroll: TIOCLINUX");
   close(fd);
 }
+
+void set_lut(void)
+{
+  int fd;
+  struct {
+    char subcode;
+    char padding[3];
+    uint32_t lut[8];
+  } l = { TIOCL_SELLOADLUT, 0, 0, 0,
+    0x00000000, /* control chars     */
+    0x03FFE000, /* digits and "-./"  */
+    0x87FFFFFE, /* uppercase and '_' */
+    0x07FFFFFE, /* lowercase         */
+    0x00000000,
+    0x00000000,
+    0xFF7FFFFF, /* latin-1 accented letters, not multiplication sign */
+    0xFF7FFFFF  /* latin-1 accented letters, not division sign */
+  }; /* all of Unicode above U+00FF is considered "word" chars, even
+        frames and the likes */
+
+  fd = open("/dev/tty0",O_RDWR);
+  if(ioctl(fd, TIOCLINUX, &l)<0)
+    perror("set_lut: TIOCLINUX");
+  close(fd);
+}
-- 
2.10.2

